### PR TITLE
`wolfSSL_PEM_write_bio_PUBKEY` needs to write only the public part

### PR DESCRIPTION
**Pull request by julek-wolfssl**
The `wolfSSL_PEM_write_bio_PUBKEY` output can't contain the private portion of the key. This output could be used to distribute the public key and if it contains the private part then it gets leaked to others.

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -34562,44 +34543,124 @@ WOLFSSL_RSA *wolfSSL_PEM_read_bio_RSA_PUBKEY(WOLFSSL_BIO* bio,WOLFSSL_RSA** rsa,
...
+            derSz = wc_DsaKeyToPublicDer((DsaKey*)key->dsa->internal, derBuf,
+                    derSz);
+            if (derSz < 0) {
+                WOLFSSL_MSG("wc_DsaKeyToDer failed");
+                break;
+            }
+            break;
+#endif
+#ifdef HAVE_ECC
+        case EVP_PKEY_EC:
```
Comment (anhu) [inline src/ssl.c:34597]: does this cover curve25519 and curve448 ?

File: src/ssl.c [hunk 2]
```diff
@@ -34608,6 +34669,16 @@ int wolfSSL_PEM_write_bio_PUBKEY(WOLFSSL_BIO* bio, WOLFSSL_EVP_PKEY* key)
     return WOLFSSL_SUCCESS;
 }
```
Comment (anhu) [inline src/ssl.c:34670]: something seems wrong with the indentation or braces.  I can't figure out the problem thought.

File: src/ssl.c [hunk 3]
```diff
@@ -34562,44 +34543,124 @@ WOLFSSL_RSA *wolfSSL_PEM_read_bio_RSA_PUBKEY(WOLFSSL_BIO* bio,WOLFSSL_RSA** rsa,
...
     if (bio == NULL || key == NULL) {
+        WOLFSSL_MSG("Bad parameters");
         return WOLFSSL_FAILURE;
     }
 
-    keyDer = (byte*)key->pkey.ptr;
+    switch (key->type) {
+#ifndef NO_RSA
+        case EVP_PKEY_RSA:
+            if ((derSz = wolfSSL_RSA_To_Der(key->rsa, &derBuf, 1)) < 0) {
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/ssl.c:34567]: `wolfSSL_RSA_To_Der` does that for me.

File: src/ssl.c (hunk 1 above)
Comment (julek-wolfssl) [reply src/ssl.c:34597]: They are not supported in the EVP layer.

File: src/ssl.c (hunk 2 above)
Comment (julek-wolfssl) [reply src/ssl.c:34670]: I think its an issue in Github's diff viewer. Its correct if you checkout this branch and look locally.

Comment (anhu) [reply src/ssl.c:34670]: Are you using tabs?

File: src/ssl.c (hunk 1 above)
Comment (anhu) [reply src/ssl.c:34597]: right....I think they were introduced into openssl's 1.1.1 release branch.

File: src/ssl.c (hunk 2 above)
Comment (julek-wolfssl) [reply src/ssl.c:34670]: Nope, eclipse is set to spaces.

File: src/ssl.c (hunk 1 above)
Comment (julek-wolfssl) [reply src/ssl.c:34597]: That might be true but we just don't support them yet. There hasn't been enough of a demand for it.

Comment (anhu) [reply src/ssl.c:34597]: Yes. My point was that I'm not familiar with what version of OpenSSL this is supposed to be compatible with.

File: src/ssl.c [hunk 4]
```diff
@@ -34562,44 +34543,125 @@ WOLFSSL_RSA *wolfSSL_PEM_read_bio_RSA_PUBKEY(WOLFSSL_BIO* bio,WOLFSSL_RSA** rsa,
 
 #endif /* defined(OPENSSL_EXTRA) || defined(OPENSSL_ALL) && !defined(NO_RSA) */
 
-/* Takes a public key and writes it out to a WOLFSSL_BIO
- * Returns WOLFSSL_SUCCESS or WOLFSSL_FAILURE
+/**
+ * The regen parameter determines if we will regenerate the DER buffer to
```
Comment (SparkiDev) [inline src/ssl.c:34547]: No longer a region parameter?

File: src/ssl.c [hunk 5]
```diff
@@ -34502,28 +34504,7 @@ int wolfSSL_PEM_write_bio_RSA_PUBKEY(WOLFSSL_BIO* bio, WOLFSSL_RSA* rsa)
...
-        WOLFSSL_MSG("wolfSSL_RSA_To_Der failed");
-        return WOLFSSL_FAILURE;
-    }
-
-    if (derBuf == NULL) {
-        WOLFSSL_MSG("wolfSSL_RSA_To_Der failed to get buffer");
-        return WOLFSSL_FAILURE;
-    }
-
-    pkey->pkey.ptr = (char*)XMALLOC(derSz, bio->heap,
```
Comment (SparkiDev) [inline src/ssl.c:34515]: DER encoding no longer cached against PKEY?

File: src/ssl.c [hunk 6]
```diff
@@ -34562,44 +34543,125 @@ WOLFSSL_RSA *wolfSSL_PEM_read_bio_RSA_PUBKEY(WOLFSSL_BIO* bio,WOLFSSL_RSA** rsa,
...
+#endif
+#if !defined(NO_DSA) && !defined(HAVE_SELFTEST) && (defined(WOLFSSL_KEY_GEN) || \
+        defined(WOLFSSL_CERT_GEN))
+        case EVP_PKEY_DSA:
+            if (key->dsa == NULL) {
+                WOLFSSL_MSG("key->dsa is null");
+                break;
+            }
+            derSz = MAX_DSA_PUBKEY_SZ;
+            /* Don't use heap hint since wolfSSL_RSA_To_Der doesn't
```
Comment (SparkiDev) [inline src/ssl.c:34581]: wolfSSL_RSA_to_Der is internal and the function signature can be changed to take heap.
Should using bio->heap and that needs to be passed in to this function.

File: src/ssl.c [hunk 7]
```diff
@@ -34562,44 +34543,125 @@ WOLFSSL_RSA *wolfSSL_PEM_read_bio_RSA_PUBKEY(WOLFSSL_BIO* bio,WOLFSSL_RSA** rsa,
...
+        XFREE(derBuf, NULL, DYNAMIC_TYPE_DER);
+        return WOLFSSL_FAILURE;
     }
 
-    ret = wc_DerToPemEx(keyDer, key->pkey_sz, tmp, pemSz,
-                        NULL, PUBLICKEY_TYPE);
+    ret = wc_DerToPem(derBuf, derSz, pemBuf, pemSz, PUBLICKEY_TYPE);
+    XFREE(derBuf, NULL, DYNAMIC_TYPE_DER);
     if (ret < 0) {
         WOLFSSL_LEAVE("wolfSSL_PEM_write_bio_PUBKEY", ret);
```
Comment (SparkiDev) [inline src/ssl.c:34658]: Wrong function name now.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/ssl.c (hunk 4 above)
Comment (julek-wolfssl) [reply src/ssl.c:34547]: Yup. Refactored to do regen inside `WriteBioPUBKEY`.

File: src/ssl.c (hunk 5 above)
Comment (julek-wolfssl) [reply src/ssl.c:34515]: `pkey` gets freed in this function anyway. No need to cache.

File: src/ssl.c (hunk 6 above)
Comment (julek-wolfssl) [reply src/ssl.c:34581]: Add heap hint argument.

File: src/ssl.c (hunk 7 above)
Comment (julek-wolfssl) [reply src/ssl.c:34658]: Corrected.

**Review by SparkiDev — APPROVED** [review]

**Pull request merged**